src/node/progress_tracker.h

```diff
@@ -0,0 +1,179 @@
+            node_id, it->second.root, signature_size, sig.data()))
+        {
+          // NOTE: We need to handle this case but for now having this make a
+          // test fail will be very handy
+          throw ccf::ccf_logic_error(fmt::format(
+            "Signature verification from {} FAILED, view:{}, seqno:{}",
+            node_id,
+            view,
+            seqno));
+          return;
+        }
+        LOG_TRACE_FMT(
+          "Signature verification from {} passed, view:{}, seqno:{}",
+          node_id,
+          view,
+          seqno);
+      }
+
+      std::vector<uint8_t> sig_vec;
+      CCF_ASSERT(
+        signature_size < sig.size(),
+        fmt::format(
+          "Invalid signature size, signature_size:{}, sig.size:{}",
+          signature_size,
+          sig.size()));
+      sig_vec.assign(sig.begin(), sig.begin() + signature_size);
+
+      auto& cert = it->second;
+      cert.sigs.insert(std::pair<kv::NodeId, std::vector<uint8_t>>(
+        node_id, std::move(sig_vec)));
+    }
+
+    void record_primary(
+      kv::Consensus::View view,
+      kv::Consensus::SeqNo seqno,
+      crypto::Sha256Hash& root)
+    {
+      auto it = certificates.find(CertKey(view, seqno));
+      if (it == certificates.end())
+      {
+        certificates.insert(std::pair<CertKey, Certificate>(
+          CertKey(view, seqno), Certificate(root)));
+        return;
+      }
+      else
+      {
+        // We received some entries before we got the root so we now need to
+        // verify the signatures
+        auto& cert = it->second;
+        for (auto& sig : cert.sigs)
+        {
+          if (!verify_signature(
+                sig.first, cert.root, sig.second.size(), sig.second.data()))
+          {
+            // NOTE: We need to handle this case but for now having this make a
+            // test fail will be very handy
+            throw ccf::ccf_logic_error(fmt::format(
+              "Signature verification from {} FAILED, view:{}, seqno:{}",
+              sig.first,
+              view,
+              seqno));
+          }
+        }
+      }
+
+      if (it->second.root != root)
+      {
+        // NOTE: At this point we have cryptographic proof that someone is being
+        // dishonest we need to work out what to do.
+        throw ccf::ccf_logic_error("We have proof someone is being dishonest");
+      }
+    }
+
+    void set_node_id(kv::NodeId id_)
+    {
+      id = id_;
+    }
+
+  private:
+    kv::NodeId id;
+    ccf::Nodes& nodes;
+
+    struct CertKey
+    {
+      CertKey(kv::Consensus::View view_, kv::Consensus::SeqNo seqno_) :
+        view(view_),
+        seqno(seqno_)
+      {}
+
+      kv::Consensus::View view;
+      kv::Consensus::SeqNo seqno;
+
+      bool operator<(const CertKey& rhs) const
+      {
+        if (seqno == rhs.seqno)
+        {
+          return view < rhs.view;
+        }
+        return seqno < rhs.seqno;
+      }
+    };
+
+    struct Certificate
```

Comment:
Minor: Is this the name used in the PBFT terminology? I wonder if this should be evidence as we've already got a few things called certificates in CCF?

Reply:
Perhaps TxCertificate, TxCert.. ?

Reply:
Sure i can use CommitCert. This is a private struct scoped within ccf::ProgressTracker so i don't think there is much chance of confusion